### DISCUSSION TITLE
Extend search

### DISCUSSION THREAD

**Discussion by helmut72** [outline/outline#4167] [Ideas]
Now it's possible to search from the beginning of a word:
aabbcc finds aabbccdd

But:
bbcc doesn't find aabbccdd

Would be cool to extend search feature to match the chars from any position. Thank you.

**Comment by almereyda** [comment]
Could this be externalised with using an optional, separate search appliance?

Meilisearch, Typesense and even ElasticSearch/OpenSearch come into mind, depending on the desired deployment environment.

**Reply by tommoor** [reply]
Yea, I've certainly considered it – to get real fulltext search that supports CJK languages it's going to be a requirement I think. The thing that really hurts from a maintenance point of view is you'd likely need to retain all the existing search code as a fallback.

**Reply by almereyda** [reply]
Some have refrained before from using Meilisearch or Typesense, because they are not v1 yet and their long-term support is unclear. From the perspective of Outline being a start-up that also runs production with a version below one, this might not be an issue.

- Meilisearch supports different language systems  
  https://docs.meilisearch.com/learn/what_is_meilisearch/language.html
  - Its [documentation announces C and J](https://docs.meilisearch.com/learn/advanced/tokenization.html) support, while [the associated tracking issue](https://github.com/meilisearch/meilisearch/issues/624) also mentions Korean (via the Lindera tokenizer).
    - Their tokenizer lives in a separate repository at https://github.com/meilisearch/charabia

Yet most often, other SaaS vendors rely on SaaS offers for such externalised services.

- https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/supported-languages/

This can be a downside for people with privacy concerns, or those intending to self-host their complete stack.

**Reply by helmut72** [reply]
In the first step it would be ok if my example works without any external search engine. I can't imagine searching for chars is limited to the beginning of chars after a space char.

**Comment by almereyda** [comment]
Meilisearch have released v1.0

- https://blog.meilisearch.com/v1-enterprise-ready-stable/

**Reply by tommoor** [reply]
Nice, I'm also investigating ML-driven search – it might make sense to just jump straight to that given the way the tech is going

**Reply by almereyda** [reply]
This message ought to be received quite clearly: _"enterprise ready stable", yes.

**Comment by helmut72** [comment]
Meilisearch would be also ok. As long as the search engine is easy to install and isn't something bloated like Elasticsearch, that explodes a Raspberry or any other low power/memory machine, every other search engine is ok.

But a built-in case insensitive search for any chars independent from space char would be great for the start.

**Reply by almereyda** [reply]
![IMG_20230211_184551](https://user-images.githubusercontent.com/1645308/218273030-f708e0e1-0c6b-4094-9dd6-9f64445b9b1e.jpg)

**Comment by jrevillard** [comment]
Hello,

Anything planned for this please ?

Best,
Jerome

**Comment by helmut72** [comment]
Since local filesystem is now supported next to Minio, I don't do only tests with Outline, but using it more serious. This was in time, I've started to test Bookstack...

Therefore a bit more extended searching becomes interesting (next to tags).

Do you have any roadmap at all?
